cmake/Hardening.cmake

```diff
@@ -0,0 +1,205 @@
+			"-mretpoline"
+			"-mspeculative-load-hardening"
+			"-lvi-load-hardening"
+			"-lvi-cfi"
+			
+			#"-fsanitize=safe-stack;compiler-rt" # https://clang.llvm.org/docs/SafeStack.html
+			"-fsanitize=address" # https://clang.llvm.org/docs/AddressSanitizer.html
+			
+			# TODO implement compiler flag dependence on libs linking
+			#"-fsanitize=undefined;ubsan" # https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html , gcc also has it
+			
+			#"-fsanitize=thread" # https://clang.llvm.org/docs/ThreadSanitizer.html , 15x slowdown and 10x memory overhead
+			#"-fsanitize=memory" # https://clang.llvm.org/docs/MemorySanitizer.html 3x slowdown
+			#"-fsanitize=dataflow" # https://clang.llvm.org/docs/DataFlowSanitizer.html, taint analysis, requires explicit annotation of code
+			
+			#"-fvtable-verify=std;vtv"
+			
+			# this conflicts with gcc which now has -fcf-protection=full hardcoded
+			"-fcf-protection=none -mindirect-branch"
+			"-fcf-protection=none -mindirect-branch=thunk-extern"
+			"-fcf-protection=none -mindirect-branch=thunk-inline"
+			"-fcf-protection=none -mindirect-return"
+			"-fcf-protection=none -mindirect-branch-register"
+			"-fcf-protection=none -mindirect-branch-loop"
+			
+			"-x86-speculative-load-hardening"
+			"-mno-indirect-branch-register"
+		)
+
+		if(HARDENING_SSE2)
+			list(APPEND HARDENING_COMPILER_FLAGS
+				"-mlfence-after-load=yes"
+				"-mlfence-before-indirect-branch=all"
+				"-mlfence-before-ret=not"
+			)
+		endif(HARDENING_SSE2)
+
+		# some flags are bugged in GCC
+		if(CMAKE_CXX_COMPILER_ID MATCHES "GNU")
+		else()
+			list(APPEND HARDENING_COMPILER_FLAGS
+				"-ftrapv" # https://gcc.gnu.org/bugzilla/show_bug.cgi?id=35412
+			)
+		endif()
+		
+		# GCC 9 has removed these flags
+		if(CMAKE_CXX_COMPILER_ID MATCHES "GNU" AND (CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 9 AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 10))
+			message(STATUS "GCC 9 removes some hardening flags but doesn't fail if they are present, instead shows deprecation message. In order to not to put garbage into warnings we don't insert them. See the code of Harden.cmake for the details.")
+		else()
+			list(APPEND HARDENING_COMPILER_FLAGS
+				"-mmitigate-rop" 
+				"-fcheck-pointer-bounds"
+				"-fchkp-treat-zero-size-reloc-as-infinite"
+				"-fchkp-first-field-has-own-bounds"
+				"-fchkp-narrow-bounds"
+				"-fchkp-narrow-to-innermost-array"
+				"-fchkp-optimize"
+				"-fchkp-use-fast-string-functions"
+				"-fchkp-use-nochk-string-functions"
+				"-fchkp-use-static-const-bounds"
+			)
+		endif()
+		
+		list(APPEND HARDENING_LINKER_FLAGS
+			"-Wl,-O1"
+			"-Wl,--sort-common"
+			"-Wl,--as-needed"
+			"-Wl,-flto"
+		)
+		if(CMAKE_SYSTEM_NAME MATCHES "Windows")
+			list(APPEND HARDENING_LINKER_FLAGS
+				"-Wl,--export-all-symbols"
+				"-Wl,--nxcompat"
+				"-Wl,--dynamicbase"
+			)
+			if(CMAKE_SIZEOF_VOID_P EQUAL 8)
+			#	list(APPEND HARDENING_LINKER_FLAGS "-Wl,--image-base,0x140000000") # doesn't work for this project
```

Comment:
Can you remove the unneeded commented out codes?

Reply:
I probably should uncomment it. It was not working for my Qt-based project, but may work on this one. Needs testing on a Windows machine.